flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/BlockingPhysicalFilePool.java

```diff
@@ -0,0 +1,96 @@
+
+import org.apache.flink.runtime.state.CheckpointedStateScope;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A Blocking {@link PhysicalFilePool} which may block when polling physical files. This class try
+ * best to reuse a physical file until its size > maxFileSize.
+ */
+public class BlockingPhysicalFilePool extends PhysicalFilePool {
+
+    private final AtomicBoolean exclusivePhysicalFilePoolInitialized;
+
+    public BlockingPhysicalFilePool(
+            long maxFileSize, PhysicalFile.PhysicalFileCreator physicalFileCreator) {
+        super(maxFileSize, physicalFileCreator);
+        this.exclusivePhysicalFilePoolInitialized = new AtomicBoolean(false);
+    }
+
+    @Override
+    public boolean tryPutFile(
+            FileMergingSnapshotManager.SubtaskKey subtaskKey, PhysicalFile physicalFile)
+            throws IOException {
+        if (physicalFile.getSize() < maxFileSize) {
+            return getFileQueue(subtaskKey, physicalFile.getScope()).offer(physicalFile);
+        } else {
+            getFileQueue(subtaskKey, physicalFile.getScope())
+                    .offer(physicalFileCreator.perform(subtaskKey, physicalFile.getScope()));
+            return false;
+        }
+    }
+
+    @Override
+    @Nonnull
+    public PhysicalFile pollFile(
+            FileMergingSnapshotManager.SubtaskKey subtaskKey, CheckpointedStateScope scope)
+            throws IOException {
+        initialize(subtaskKey, scope);
+        try {
+            return ((BlockingQueue<PhysicalFile>) getFileQueue(subtaskKey, scope)).take();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void initialize(
+            FileMergingSnapshotManager.SubtaskKey subtaskKey, CheckpointedStateScope scope)
+            throws IOException {
+        if (scope.equals(CheckpointedStateScope.SHARED)) {
+            AtomicBoolean init = new AtomicBoolean(false);
+            Queue<PhysicalFile> fileQueue =
+                    sharedPhysicalFilePoolBySubtask.computeIfAbsent(
+                            subtaskKey,
+                            key -> {
+                                init.set(true);
+                                return createFileQueue();
+                            });
+            if (init.get()) {
+                fileQueue.offer(physicalFileCreator.perform(subtaskKey, scope));
+            }
+        } else if (scope.equals(CheckpointedStateScope.EXCLUSIVE)
+                && exclusivePhysicalFilePoolInitialized.compareAndSet(false, true)) {
+            getFileQueue(subtaskKey, scope).offer(physicalFileCreator.perform(subtaskKey, scope));
+        }
+    }
+
+    @Override
+    protected Queue<PhysicalFile> createFileQueue() {
+        return new LinkedBlockingQueue<>();
```

Comment:
I noticed that in `NonBlockingPhysicalFilePool` file queue is `ConcurrentLinkedQueue`. Do both `NonBlockingPhysicalFilePool ` and `BlockingPhysicalFilePool ` should be thread-safe?

Reply:
Yes, you're right.
As you could see the comment in `PhysicalFilePool`.